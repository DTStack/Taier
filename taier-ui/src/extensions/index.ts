/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { IExtension } from '@dtinsight/molecule/esm/model';
import InitializeExtension from './init';
import CatalogueExtension from './catalogue';
import PanelExtension from './panel';
import EditorExtension from './editor';
import FolderTreeExtension from './folderTree';
import { ExtendsSparkSQL } from './languages';
import MenuExtension from './menu';
import DTStackTheme from './dtstackTheme';
import { LocaleExtension } from './i18n';
import ActionExtension from './action';
import ColorThemeExtensions from './themes';
import AuxiliaryBarExtensions from './auxiliaryBar';
import ConfirmExtension from './confirm';

export const extensions: IExtension[] = [
    LocaleExtension,
    DTStackTheme,
    new ConfirmExtension(),
    new ColorThemeExtensions(),
    new ActionExtension(),
    new InitializeExtension(),
    new PanelExtension(),
    new CatalogueExtension(),
    new EditorExtension(),
    new FolderTreeExtension(),
    new ExtendsSparkSQL(),
    new MenuExtension(),
    new AuxiliaryBarExtensions(),
];
