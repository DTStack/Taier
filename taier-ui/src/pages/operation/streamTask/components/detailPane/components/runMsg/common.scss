.vertex-diagram {
    background: #fff;
    width: 280px;
    user-select: none;

    .vertex-title {
        width: 100%;
        display: grid;
        grid-template-columns: auto 32px;
        align-items: center;
        height: 36px;
        line-height: 36px;
        border-radius: 4px 4px 0 0;
        color: #fff;
        padding: 0 0 0 12px;

        .t_title {
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
            float: left;
        }
    }

    .vertex-content {
        padding-right: 36px;
        font-size: 12px;
        font-family: PingFangSC-Regular, "PingFang SC";
        font-weight: 400;
        color: rgba(102, 102, 102, 1);
        height: 198px;

        .tcolumn {
            overflow: hidden;
            text-overflow: ellipsis;
            display: -webkit-box;
            -webkit-line-clamp: 2;
            -webkit-box-orient: vertical;
            padding: 12px 15px 0;
            margin-bottom: 8px;
            text-align: left;
        }

        .t-text-col {
            height: 30px;
            line-height: 30px;

            .t-text-col-key {
                float: left;
                text-indent: 12px;

                .t-text-col_img {
                    width: 12px;
                    height: 12px;
                    margin: -2px 8px 0 0;
                }
            }

            .t-text-col-value {
                float: right;
                font-family: PingFangSC-Medium, "PingFang SC";
                font-weight: 500;
                color: rgba(51, 51, 51, 1);
            }
        }
    }
}
