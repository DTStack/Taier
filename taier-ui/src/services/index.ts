import { container } from 'tsyringe';
import EditorActionBarService from './editorActionBarService';
import ExecuteService from './executeService';
import CatalogueService from './catalogueService';
import BreadcrumbService from './breadcrumbService';
import RightBarService from './rightBarService';
import TaskRenderService from './taskRenderService';
import DataSourceService from './dataSourceService';
import TaskParamsService from './taskParamsService';

const editorActionBarService = container.resolve(EditorActionBarService);
const executeService = container.resolve(ExecuteService);
const catalogueService = container.resolve(CatalogueService);
const breadcrumbService = container.resolve(BreadcrumbService);
const rightBarService = container.resolve(RightBarService);
const taskRenderService = container.resolve(TaskRenderService);
const dataSourceService = container.resolve(DataSourceService);
const taskParamsService = container.resolve(TaskParamsService);

export {
    editorActionBarService,
    catalogueService,
    executeService,
    breadcrumbService,
    rightBarService,
    taskRenderService,
    dataSourceService,
    taskParamsService,
};
