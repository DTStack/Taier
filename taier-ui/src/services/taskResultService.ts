/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@dtinsight/molecule/esm/react';
import moment from 'moment';

export interface ITaskResultService {
    /**
     * 增量添加日志信息
     *
     * 自动添加换行标识符
     */
    appendLogs: (key: string, log: string) => void;
    /**
     * 清除任务执行的日志信息
     */
    clearLogs: (key: string) => void;
    /**
     * 任务执行结果
     */
    setResult: (key: `${string}-${string}`, result: any) => void;
    /**
     * 清除执行结果
     */
    clearResult: (key: string) => void;
}

/**
 * 输出固定格式的日志前缀
 * @example `[14:00:01] <error> xxxx`
 */
export function createLog(log: string, type = 'info') {
    return `[${moment().format('HH:mm:ss')}] <${type}> ${log}`;
}

/**
 * 生成固定格式的 title
 * @example `=====title=====`
 * @example `===============`
 */
export function createTitle(title = '') {
    const baseLength = 15;
    const offsetLength = Math.floor((1.5 * title.length) / 2);
    const arr = new Array(Math.max(baseLength - offsetLength, 5));
    const wraptext = arr.join('=');
    return `${wraptext}${title}${wraptext}`;
}

export function createLinkMark(attrs: Record<string, string | null>) {
    return `${window.location.origin}${attrs.href}`;
}
export interface ITaskResultStates {
    /**
     * 存储不同任务执行的日志结果
     */
    logs: Record<string, string>;
    /**
     * 任务执行结果
     * @key 前者为 tabId，后者为基于 sqlText 生成的 md5 码
     */
    results: Record<`${string}-${string}`, string[][]>;
}

class TaskResultService extends Component<ITaskResultStates> implements ITaskResultService {
    protected state: ITaskResultStates;

    constructor() {
        super();
        this.state = {
            logs: {},
            results: {},
        };
    }

    public setResult(key: `${string}-${string}`, res: any) {
        const nextResults = this.state.results;
        nextResults[key] = res;
        this.setState({ results: { ...nextResults } });
    }

    public clearResult(key: string) {
        const nextResults = this.state.results;
        Reflect.deleteProperty(nextResults, key);
        this.setState({ results: { ...nextResults } });
    }

    public appendLogs(key: string, log: string) {
        const nextLogs = this.state.logs;
        nextLogs[key] = nextLogs[key] || '';
        nextLogs[key] += `${log}\n`;
        this.setState({
            logs: { ...nextLogs },
        });
    }

    public clearLogs(key: string) {
        const nextLogs = this.state.logs;
        nextLogs[key] = '';
        this.setState({
            logs: { ...nextLogs },
        });
    }
}

/**
 * 处理任务执行结果的 service
 */
export default new TaskResultService();
