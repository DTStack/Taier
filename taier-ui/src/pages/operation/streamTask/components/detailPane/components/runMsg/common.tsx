import { useState, useEffect, useMemo } from 'react';
import type { mxCell, mxGraph } from 'mxgraph';
import * as ReactDOMServer from 'react-dom/server';
import { Modal } from 'antd';
import MxGraphContainer from '@/components/mxGraph/container';
import { cloneDeep } from 'lodash';
import type { IFlinkJsonProps } from '.';
import classnames from 'classnames';
import './common.scss';

interface ICommonProps {
    flinkJson: IFlinkJsonProps[];
    loading: boolean;
    refresh: () => void;
}

type ISubJobVertices = IFlinkJsonProps['subJobVertices'][number];

interface ITopological {
    id: string;
    title: string;
    desc: string;
    delay: number;
    parallelism: number;
    received: number;
    sent: number;
    backPressured: number;
    subJobVertices?: ISubJobVertices[];
    childNode: ITopological[];
}

const getBackPressureColor = (backPressure: number, type: string) => {
    let color = '';
    if (backPressure >= 0 && backPressure <= 0.1) {
        // 正常
        if (type === 'title') {
            color = 'linear-gradient(298deg,rgba(90,203,255,1) 0%,rgba(36,145,247,1) 100%)';
        } else {
            color = '#F1FAFF';
        }
    } else if (backPressure >= 0.1 && backPressure <= 0.5) {
        // 低反压
        if (type === 'title') {
            color = 'linear-gradient(270deg,rgba(255,190,76,1) 0%,rgba(255,160,41,1) 100%)';
        } else {
            color = '#fff9f0';
        }
    } else if (backPressure > 0.5 && backPressure <= 1) {
        // 高反压
        if (type === 'title') {
            color = 'linear-gradient(270deg,#FF5F5C 0%,#ea8785 100%)';
        } else {
            color = '#ff5f5c4d';
        }
    }
    return color;
};

export default function Common({ flinkJson, loading, refresh }: ICommonProps) {
    const [graphData, setGraphData] = useState<ITopological[] | null>(null);
    // 当前选中的 cell
    const [selectedCell, setSelectedCell] = useState<mxCell | null>(null);
    const [visible, setVisible] = useState(false);

    const handleClick = (cell: mxCell, graph: mxGraph, event: React.MouseEvent<HTMLElement, MouseEvent>) => {
        if ((event.target as HTMLImageElement).nodeName === 'IMG') {
            if ((cell.value as ITopological)?.subJobVertices) {
                setVisible(true);

                setSelectedCell(cell);
            }
        }
    };

    const handleRenderCell = (cell: mxCell) => {
        if (cell.value) {
            const { backPressured, title, desc, delay, parallelism, received, sent, subJobVertices } =
                cell.value as ITopological;
            const res = ReactDOMServer.renderToString(
                <div className="vertex-diagram" style={{ background: getBackPressureColor(backPressured, 'title') }}>
                    <div className="vertex-title">
                        <div className="t_title">{title}</div>
                        {subJobVertices?.length && <img className="cursor-pointer" src="images/expand.svg" />}
                    </div>
                    <div
                        className="vertex-content"
                        style={{ background: getBackPressureColor(backPressured, 'content') }}
                    >
                        <div className="tcolumn" title={desc}>
                            {desc}
                        </div>
                        {[
                            { field: 'delay', title: 'Delay', data: `${delay}ms` },
                            { field: 'parallelism', title: 'Parallelism', data: parallelism },
                            { field: 'received', title: 'Record Received', data: received },
                            { field: 'sent', title: 'Record Sent', data: sent },
                            {
                                field: 'dashboard',
                                title: 'BackPressured(max)',
                                data: `${(backPressured * 100).toFixed(0)}%`,
                            },
                        ].map(({ field, title: fieldTitle, data }) => (
                            <div className="t-text-col" title={fieldTitle} key={field}>
                                <span className="t-text-col-key">
                                    <img src={`images/${field}.svg`} className="t-text-col_img" />${fieldTitle}
                                </span>
                                <span className="t-text-col-value">{data}</span>
                            </div>
                        ))}
                    </div>
                </div>
            );
            return res;
        }
        return '';
    };

    const convertArrToLinkedList = (arr?: ISubJobVertices[]): ITopological[] | null => {
        let res: ITopological[] | null = null;
        arr?.forEach(
            function (this: { point: ITopological[] | null }, vertice, index) {
                if (!this.point) {
                    res = [
                        {
                            id: vertice.id,
                            title: `Operator ${index}`,
                            desc: vertice.name,
                            delay: Object.values(vertice.delayMapList)[0],
                            parallelism: vertice.parallelism,
                            received: Object.values(vertice.recordsReceivedMap)[0],
                            sent: Object.values(vertice.recordsSentMap)[0],
                            // 子节点的最大反压数据从父节点获取
                            backPressured: selectedCell?.value?.backPressured || 0,
                            childNode: [],
                        },
                    ];
                    this.point = res[0].childNode;
                } else {
                    this.point[0] = {
                        id: vertice.id,
                        title: `Operator ${index}`,
                        desc: vertice.name,
                        delay: Object.values(vertice.delayMapList)[0],
                        parallelism: vertice.parallelism,
                        received: Object.values(vertice.recordsReceivedMap)[0],
                        sent: Object.values(vertice.recordsSentMap)[0],
                        backPressured: selectedCell?.value?.backPressured || 0,
                        childNode: [],
                    };
                    this.point = this.point[0].childNode;
                }
            },
            { point: res }
        );

        return res;
    };

    const handleRefreshSub = () => {
        setSelectedCell((c) => cloneDeep(c));
    };

    useEffect(() => {
        if (flinkJson.length) {
            setGraphData(
                flinkJson.map((json) => {
                    return {
                        id: json.jobVertexId,
                        title: `Chain ${json.subJobVertices.length} Operators`,
                        desc: json.jobVertexName,
                        delay: Number(Object.values(json.delayMap)[0]) || 0,
                        parallelism: json.parallelism,
                        received: json.recordsReceived,
                        sent: json.recordsSent,
                        backPressured: Math.max(...Object.values(json.backPressureMap)),
                        subJobVertices: json.subJobVertices,
                        childNode: [],
                    };
                })
            );
        }
    }, [flinkJson]);

    const subGraphData = useMemo(() => {
        if (selectedCell) {
            return convertArrToLinkedList((selectedCell.value as ITopological)?.subJobVertices);
        }
        return null;
    }, [selectedCell]);

    if (!graphData) {
        return (
            <div className={classnames('full-w', 'flex-1', 'flex', 'items-center', 'justify-center')}>
                暂未生成拓扑图
            </div>
        );
    }

    return (
        <>
            <MxGraphContainer<ITopological>
                config={{
                    tooltips: false,
                }}
                vertexSize={{
                    width: 280,
                    height: 234,
                }}
                onClick={handleClick}
                loading={loading}
                graphData={graphData}
                onRenderCell={handleRenderCell}
                onRefresh={refresh}
            />
            <Modal
                visible={visible}
                title="工作流"
                onCancel={() => setVisible(false)}
                footer={null}
                zIndex={1000}
                width={900}
                destroyOnClose
            >
                <div style={{ height: 500 }}>
                    <MxGraphContainer<ITopological>
                        config={{
                            tooltips: false,
                        }}
                        vertexSize={{
                            width: 280,
                            height: 234,
                        }}
                        direction="west"
                        onClick={handleClick}
                        loading={loading}
                        graphData={subGraphData}
                        onRenderCell={handleRenderCell}
                        onRefresh={handleRefreshSub}
                    />
                </div>
            </Modal>
        </>
    );
}
