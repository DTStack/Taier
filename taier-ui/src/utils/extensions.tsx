/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import molecule from '@dtinsight/molecule';
import type { IFolderTreeNodeProps } from '@dtinsight/molecule/esm/model';
import { SyntaxIcon, ResourceIcon } from '@/components/icon';
import type { RESOURCE_TYPE } from '@/constant';
import { ID_COLLECTIONS, CATALOGUE_TYPE, TASK_TYPE_ENUM } from '@/constant';
import type { CatalogueDataProps, IOfflineTaskProps } from '@/interface';
import { IJobType } from '@/interface';
import { executeService, taskRenderService } from '@/services';
import taskResultService, { createLog } from '@/services/taskResultService';
import Result from '@/components/task/result';
import { filterSql } from '.';
import stream from '@/api';
import { TreeViewUtil } from '@dtinsight/molecule/esm/common/treeUtil';
import taskSaveService from '@/services/taskSaveService';
import md5 from 'md5';
import { EventBus } from '@dtinsight/molecule/esm/common/event';

/**
 * 根据不同任务渲染不同的图标
 */
export function fileIcon(type: TASK_TYPE_ENUM | RESOURCE_TYPE | null, source: CATALOGUE_TYPE): string | JSX.Element {
    switch (source) {
        case CATALOGUE_TYPE.TASK: {
            return taskRenderService.renderTaskIcon(type as TASK_TYPE_ENUM);
        }
        case CATALOGUE_TYPE.RESOURCE: {
            return <ResourceIcon style={{ color: '#0065f6' }} />;
        }
        case CATALOGUE_TYPE.FUNCTION:
        default:
            return 'code';
    }
}

/**
 * 运行任务
 */
export function runTask(current: molecule.model.IEditorGroup) {
    const currentTabData: (CatalogueDataProps & IOfflineTaskProps & { value?: string }) | undefined = current.tab?.data;
    if (currentTabData) {
        // active 日志 窗口
        const { data } = molecule.panel.getState();
        const {
            panel: { hidden },
        } = molecule.layout.getState();
        if (hidden) {
            molecule.layout.togglePanelVisibility();
        }
        molecule.panel.setState({
            current: data?.find((item) => item.id === ID_COLLECTIONS.OUTPUT_LOG_ID),
        });

        if (currentTabData.taskType === TASK_TYPE_ENUM.SYNC) {
            const params: any = {
                taskId: currentTabData.id,
                name: currentTabData.name,
                taskParams: currentTabData.taskParams,
            };
            executeService.execDataSync(currentTabData.id, params);
        } else {
            const params = {
                taskVariables: currentTabData.taskVariables || [],
                // 是否为单 session 模式, 为 true 时，支持batchSession 时，则支持批量SQL，false 则相反
                singleSession: false,
                taskParams: currentTabData.taskParams,
            };

            const value = currentTabData.value || '';
            // 需要被执行的 sql 语句
            const sqls: string[] = [];

            const field = taskRenderService.getField(currentTabData.taskType);
            if (
                field?.jobType === IJobType.SQL &&
                taskRenderService.getRenderKind(currentTabData.taskType) === 'editor'
            ) {
                const rawSelections = molecule.editor.editorInstance.getSelections() || [];
                // 排除鼠标 focus 在 editor 中的情况
                const selections = rawSelections.filter(
                    (s) => s.startLineNumber !== s.endLineNumber || s.startColumn !== s.endColumn
                );
                // 如果存在选中行，则执行选中行
                if (selections?.length) {
                    selections?.forEach((s) => {
                        const text = molecule.editor.editorInstance.getModel()?.getValueInRange(s);
                        if (text) {
                            sqls.push(...filterSql(text));
                        }
                    });
                } else {
                    sqls.push(...filterSql(value));
                }
            } else {
                sqls.push(value);
            }

            executeService.execSql(currentTabData.id, currentTabData, params, sqls).then(() => {
                const { results } = taskResultService.getState();
                let nextActivePanel: string | null = null;
                Object.entries(results).forEach(([key, values]) => {
                    const panel = molecule.panel.getPanel(key);
                    const renderPane = () => (
                        <Result
                            data={values}
                            tab={{
                                tableType: 0,
                            }}
                            extraView={null}
                        />
                    );
                    if (!panel) {
                        const panels = molecule.panel.getState().data || [];
                        const resultPanles = panels.filter((p) => p.name?.includes('结果'));
                        const lastIndexOf = Number(resultPanles[resultPanles.length - 1]?.name?.slice(2) || '');

                        nextActivePanel = key;
                        molecule.panel.add({
                            id: key,
                            name: `结果 ${lastIndexOf + 1}`,
                            closable: true,
                            renderPane,
                        });
                    } else {
                        // 更新已有的panel
                        molecule.panel.update({
                            id: key,
                            renderPane,
                        });
                    }
                });

                const nextKey = nextActivePanel || `${currentTabData.id}-${md5(sqls.at(-1) || 'sync')}`;

                if (molecule.panel.getPanel(nextKey)) {
                    molecule.panel.setActive(nextKey);
                }
            });
        }
    }
}

/**
 * 语法检查
 */
export function syntaxValidate(current: molecule.model.IEditorGroup) {
    const currentTabData: IOfflineTaskProps | undefined = current.tab?.data;
    if (!currentTabData) return;
    // 禁用语法检查
    molecule.editor.updateActions([
        {
            id: ID_COLLECTIONS.TASK_SYNTAX_ID,
            icon: 'loading~spin',
            disabled: true,
        },
    ]);

    // active 日志 窗口
    const { data } = molecule.panel.getState();
    const {
        panel: { hidden },
    } = molecule.layout.getState();
    if (hidden) {
        molecule.layout.togglePanelVisibility();
    }
    molecule.panel.setState({
        current: data?.find((item) => item.id === ID_COLLECTIONS.OUTPUT_LOG_ID),
    });

    const logId = currentTabData.id.toString();
    taskResultService.clearLogs(logId);
    taskResultService.appendLogs(logId, createLog('语法检查开始', 'info'));

    const params = taskSaveService.transformTabDataToParams(currentTabData);

    let isSuccess = false;
    stream
        .checkSyntax(params)
        .then((res) => {
            if (res.message) {
                taskResultService.appendLogs(logId, createLog(res.message, 'error'));
            }
            if (res && res.code === 1) {
                if (res.data.code === 1) {
                    taskResultService.appendLogs(logId, createLog('语法检查通过', 'info'));
                    isSuccess = true;
                } else {
                    taskResultService.appendLogs(logId, createLog(res.data.errorMsg, 'error'));
                }
            }
        })
        .catch((e) => {
            console.trace(e);
        })
        .finally(() => {
            if (!isSuccess) {
                taskResultService.appendLogs(logId, createLog('语法检查失败！', 'error'));
            }
            // 恢复语法检查按钮
            molecule.editor.updateActions([
                {
                    id: ID_COLLECTIONS.TASK_SYNTAX_ID,
                    icon: <SyntaxIcon />,
                    disabled: false,
                },
            ]);
        });
}

export function getParentNode(treeList: IFolderTreeNodeProps, currentNode: IFolderTreeNodeProps) {
    const treeView = new TreeViewUtil(treeList);
    const parentNode = treeView.getHashMap(currentNode.id)?.parent;
    if (parentNode) {
        return treeView.getNode(parentNode);
    }
    return null;
}

/**
 * This function for executing func after switching task
 */
export function onTaskSwitch(func: () => void) {
    EventBus.subscribe(ID_COLLECTIONS.TASK_SWITCH_EVENT, () => {
        func();
    });
}
