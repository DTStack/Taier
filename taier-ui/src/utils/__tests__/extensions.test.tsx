import api from '@/api';
import { CATALOGUE_TYPE, TASK_TYPE_ENUM } from '@/constant';
import { executeService, taskRenderService } from '@/services';
import taskResultService, { createLog } from '@/services/taskResultService';
import molecule from '@dtinsight/molecule';
import { TreeViewUtil } from '@dtinsight/molecule/esm/common/treeUtil';
import { fileIcon, getParentNode, runTask, syntaxValidate } from '../extensions';

jest.mock('@/services/taskSaveService', () => {
    return {
        transformTabDataToParams: jest.fn(),
    };
});

jest.mock('@/services/taskResultService', () => {
    return {
        getState: jest.fn(() => ({ results: { test: 'test' } })),
        clearLogs: jest.fn(),
        appendLogs: jest.fn(),
        createLog: jest.fn(),
    };
});

jest.mock('@/services', () => {
    return {
        executeService: {
            execDataSync: jest.fn(),
            execSql: jest.fn(() => Promise.resolve()),
        },
        taskRenderService: {
            renderTaskIcon: jest.fn((type) => type),
            getField: jest.fn(),
            getRenderKind: jest.fn(),
        },
    };
});

jest.mock('@/api', () => {
    return {
        checkSyntax: jest.fn(),
    };
});

describe('utils/extensions', () => {
    beforeAll(() => {
        (molecule.panel.getState as jest.Mock)
            .mockReset()
            .mockImplementation(() => ({ data: [{ id: 'panel.output.log' }] }));
    });

    it('Should render file icon', () => {
        expect(fileIcon(null, CATALOGUE_TYPE.FUNCTION)).toBe('code');
        expect(fileIcon(null, CATALOGUE_TYPE.RESOURCE)).toMatchSnapshot();
        expect(fileIcon(TASK_TYPE_ENUM.SQL, CATALOGUE_TYPE.TASK)).toBe(TASK_TYPE_ENUM.SQL);

        expect(taskRenderService.renderTaskIcon).toBeCalled();
    });

    it('Should run task with some effects', async () => {
        // Mock the molecule.layout.getState
        (molecule.layout.getState as jest.Mock)
            .mockReset()
            .mockImplementationOnce(() => ({ panel: { hidden: false } }))
            .mockImplementation(() => ({ panel: { hidden: true } }));

        (molecule.panel.getPanel as jest.Mock)
            .mockReset()
            .mockImplementationOnce(() => undefined)
            .mockImplementation(() => ({ id: 'test' }));

        (taskRenderService.getField as jest.Mock).mockReset().mockImplementation(() => ({
            key: 0,
            value: 'SparkSQL',
            computeType: 1,
            jobType: 0,
        }));

        (taskRenderService.getRenderKind as jest.Mock).mockReset().mockImplementation(() => 'editor');

        runTask({
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                    name: 'test',
                    taskParams: [],
                    taskType: TASK_TYPE_ENUM.SYNC,
                },
            },
        });

        // At first time called, the panel is visible because of the return value of molecule.layout.getState
        expect(molecule.layout.togglePanelVisibility).not.toBeCalled();
        // Whatever the log panel is hidden or visible, the activated panel must be the log panel
        expect(molecule.panel.setState).toBeCalledWith({ current: { id: 'panel.output.log' } });

        expect(executeService.execDataSync).toBeCalledWith('test', {
            taskId: 'test',
            name: 'test',
            taskParams: [],
        });

        const current = {
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                    name: 'test',
                    taskParams: [],
                    taskVariables: [],
                    taskType: TASK_TYPE_ENUM.SPARK_SQL,
                    value: 'show tables; select * from A;',
                },
            },
        };
        // First time the getSelection returns undefined, so execute all sqls
        runTask(current);

        // Toggle the log panel and set this panel active since this panel is hidden
        expect(molecule.layout.togglePanelVisibility).toBeCalled();

        expect(executeService.execSql).toBeCalledWith(
            'test',
            current.tab.data,
            {
                taskVariables: [],
                singleSession: false,
                taskParams: [],
            },
            ['show tables', 'select * from A']
        );
        // Delay expect because then is a micro task
        await new Promise<void>((resolve) => {
            setTimeout(() => {
                expect(molecule.panel.add).toBeCalledWith(
                    expect.objectContaining({
                        id: 'test',
                        name: '结果 1',
                        closable: true,
                    })
                );
                expect((molecule.panel.add as jest.Mock).mock.calls[0][0].renderPane()).toMatchSnapshot();
                resolve();
            }, 0);
        });

        // Second time the getSelection returns same point, so execute all sqls
        runTask(current);
        expect(executeService.execSql).toBeCalledWith(
            'test',
            current.tab.data,
            {
                taskVariables: [],
                singleSession: false,
                taskParams: [],
            },
            ['show tables', 'select * from A']
        );

        // Third time the getSelection returns correct selection, so execute selected sql
        runTask(current);
        expect(executeService.execSql).toBeCalledWith(
            'test',
            current.tab.data,
            {
                taskVariables: [],
                singleSession: false,
                taskParams: [],
            },
            ['show tables']
        );
    });

    it('Should validate syntax failed', async () => {
        (molecule.layout.togglePanelVisibility as jest.Mock).mockClear();

        (molecule.layout.getState as jest.Mock).mockReset().mockImplementation(() => ({ panel: { hidden: false } }));

        (api.checkSyntax as jest.Mock).mockReset().mockImplementation(() => Promise.reject());

        const originalTrace = console.trace;
        console.trace = jest.fn();

        syntaxValidate({
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                },
            },
        });

        // To disabled syntax button
        expect(molecule.editor.updateActions).toBeCalledWith([
            {
                id: 'task.syntax',
                icon: 'loading~spin',
                disabled: true,
            },
        ]);

        // At first time called, the molecule.layout.getState returns false
        expect(molecule.layout.togglePanelVisibility).not.toBeCalled();
        // Whatever the log panel is hidden or visible, the activated panel must be the log panel
        expect(molecule.panel.setState).toBeCalledWith({ current: { id: 'panel.output.log' } });

        await new Promise<void>((resolve) => {
            setTimeout(() => {
                // Reject would call console.trace
                expect(console.trace).toBeCalled();

                expect(createLog).toBeCalledWith('语法检查失败！', 'error');
                expect(taskResultService.appendLogs).toBeCalled();

                // To restore the button
                expect(molecule.editor.updateActions).toBeCalledWith([
                    expect.objectContaining({
                        id: 'task.syntax',
                        disabled: false,
                    }),
                ]);

                console.trace = originalTrace;
                resolve();
            }, 0);
        });
    });

    it('Should validate syntax success', async () => {
        (molecule.layout.togglePanelVisibility as jest.Mock).mockClear();

        (molecule.layout.getState as jest.Mock).mockReset().mockImplementation(() => ({ panel: { hidden: true } }));

        (api.checkSyntax as jest.Mock)
            .mockReset()
            .mockImplementationOnce(() => Promise.resolve({ code: 100, message: 'test' }))
            .mockImplementationOnce(() => Promise.resolve({ code: 1, data: { code: 100, errorMsg: 'errorMsg' } }))
            .mockImplementation(() => Promise.resolve({ code: 1, data: { code: 1 } }));

        (createLog as jest.Mock).mockClear();
        (taskResultService.appendLogs as jest.Mock).mockClear();

        syntaxValidate({
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                },
            },
        });

        expect(molecule.layout.togglePanelVisibility).toBeCalled();

        await new Promise<void>((resolve) => {
            setTimeout(() => {
                expect(createLog).toBeCalledWith('test', 'error');
                expect(taskResultService.appendLogs).toBeCalled();

                expect(createLog).toBeCalledWith('语法检查失败！', 'error');
                expect(taskResultService.appendLogs).toBeCalled();
                resolve();
            }, 0);
        });

        syntaxValidate({
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                },
            },
        });

        await new Promise<void>((resolve) => {
            setTimeout(() => {
                expect(createLog).toBeCalledWith('errorMsg', 'error');
                expect(taskResultService.appendLogs).toBeCalled();

                expect(createLog).toBeCalledWith('语法检查失败！', 'error');
                expect(taskResultService.appendLogs).toBeCalled();
                resolve();
            }, 0);
        });

        // reset the count of log functions
        (createLog as jest.Mock).mockClear();
        (taskResultService.appendLogs as jest.Mock).mockClear();

        // succeed
        syntaxValidate({
            id: 1,
            tab: {
                id: 1,
                data: {
                    id: 'test',
                },
            },
        });

        await new Promise<void>((resolve) => {
            setTimeout(() => {
                expect(createLog).toBeCalledWith('语法检查通过', 'info');
                expect(taskResultService.appendLogs).toBeCalledTimes(2);

                expect(createLog).not.toBeCalledWith('语法检查失败！', 'error');
                expect(taskResultService.appendLogs).toBeCalledTimes(2);
                resolve();
            }, 0);
        });
    });

    it('Should support to find parent node', () => {
        const mockNode = jest.fn();
        (TreeViewUtil as jest.Mock)
            .mockImplementationOnce(() => ({
                getHashMap: jest.fn(),
                getNode: jest.fn((node) => node),
            }))
            .mockImplementation(() => ({
                getHashMap: jest.fn().mockImplementation(() => ({
                    parent: mockNode,
                })),
                getNode: jest.fn((node) => node),
            }));

        expect(
            getParentNode(
                {
                    id: '1',
                },
                {
                    id: '2',
                }
            )
        ).toBe(null);

        expect(
            getParentNode(
                {
                    id: '1',
                },
                {
                    id: '1',
                }
            )
        ).toBe(mockNode);
    });
});
