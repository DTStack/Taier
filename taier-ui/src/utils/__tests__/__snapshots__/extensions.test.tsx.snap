// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/extensions Should render file icon 1`] = `
<ResourceIcon
  style={
    Object {
      "color": "#0065f6",
    }
  }
/>
`;

exports[`utils/extensions Should run task with some effects 1`] = `
<Result
  data="test"
  extraView={null}
  tab={
    Object {
      "tableType": 0,
    }
  }
/>
`;
